File: stayhub/hasura/scripts/apply.py
"""Push metadata.py into a running Hasura.

    python -m scripts.apply

Uses the metadata API's `replace_metadata` — the whole document at once, so what runs is exactly
what is in the file. Applying table-by-table would leave whatever a previous run created and make
the file a wish rather than a description.
"""

import json
import os
import sys
import urllib.error
import urllib.request

from metadata import build

HASURA_URL = os.getenv("HASURA_URL", "http://localhost:8081")
ADMIN_SECRET = os.getenv("HASURA_ADMIN_SECRET", "stayhub-admin-secret")


def call(payload: dict) -> dict:
    request = urllib.request.Request(
        f"{HASURA_URL}/v1/metadata",
        data=json.dumps(payload).encode(),
        headers={
            "Content-Type": "application/json",
            # The admin secret bypasses every permission rule. It belongs in scripts and in the
            # console — never in a frontend, where it would hand every visitor the admin role.
            "x-hasura-admin-secret": ADMIN_SECRET,
        },
    )
    try:
        with urllib.request.urlopen(request, timeout=30) as response:
            return json.loads(response.read())
    except urllib.error.HTTPError as exc:
        body = exc.read().decode()
        print(f"Hasura rejected the request ({exc.code}):", file=sys.stderr)
        try:
            print(json.dumps(json.loads(body), indent=2), file=sys.stderr)
        except json.JSONDecodeError:
            print(body, file=sys.stderr)
        raise SystemExit(1) from exc


def main() -> None:
    print(f"Applying metadata to {HASURA_URL} …")
    call({
        "type": "replace_metadata",
        "args": {
            # Refuse rather than silently drop anything metadata.py does not mention. If a table
            # was tracked by hand in the console, this makes that visible instead of erasing it.
            "allow_inconsistent_metadata": False,
            "metadata": build(),
        },
    })

    # `replace_metadata` returning 200 only means Hasura ACCEPTED the document. Asking for the
    # inconsistency list is what confirms every table and relationship actually resolved.
    inconsistencies = call({"type": "get_inconsistent_metadata", "args": {}})
    problems = inconsistencies.get("inconsistent_objects", [])
    if problems:
        print(f"\n{len(problems)} inconsistent object(s):", file=sys.stderr)
        for item in problems:
            print(f"  - {item.get('reason')}", file=sys.stderr)
        raise SystemExit(1)

    doc = build()
    tables = doc["sources"][0]["tables"]
    print(f"  {len(tables)} tables tracked, metadata consistent")
    print()
    print("  ⚠️ No insert/update/delete permissions exist for ANY role — deliberately.")
    print("     Every write goes through FastAPI, which is what makes server-side pricing,")
    print("     the availability check and the cancellation rule impossible to bypass.")
    print()
    print(f"  Console: {HASURA_URL}  (admin secret: {ADMIN_SECRET})")


if __name__ == "__main__":
    main()

File: contractor/hasura/apply.mjs
/**
 * Pushes `metadata.mjs` into a running Hasura.
 *
 *   node hasura/apply.mjs           # apply
 *   node hasura/apply.mjs --check   # show what Hasura currently has, change nothing
 *
 * No dependencies and no Hasura CLI — Node 18+ has `fetch`, and the metadata API is one endpoint.
 * The CLI is the right tool on a real project (it does migrations and versioned metadata); here it
 * would be a binary to install before the demo runs.
 *
 * ⚠️ `replace_metadata` is a REPLACE, not a merge. Anything tracked in the console and not written
 * in metadata.mjs disappears when this runs. That is the intended behaviour — the file is the
 * source of truth, and a console click that survives a re-apply is a permission nobody reviewed.
 */

import { metadata } from './metadata.mjs'

const ENDPOINT = process.env.HASURA_ENDPOINT ?? 'http://localhost:8083'
// The admin secret bypasses every permission in metadata.mjs. It belongs in a script like this and
// in the console, and nowhere near a browser bundle.
const ADMIN_SECRET = process.env.HASURA_ADMIN_SECRET ?? 'contractor-admin-secret'

async function callMetadataApi(body) {
  const response = await fetch(`${ENDPOINT}/v1/metadata`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-hasura-admin-secret': ADMIN_SECRET,
    },
    body: JSON.stringify(body),
  })

  const text = await response.text()
  let parsed
  try {
    parsed = JSON.parse(text)
  } catch {
    // A non-JSON body here is nearly always an HTML error page from something that is not Hasura —
    // usually the wrong port. Saying so beats a JSON parse error.
    throw new Error(`${ENDPOINT} did not return JSON. Is Hasura running there?\n${text.slice(0, 300)}`)
  }

  if (!response.ok) {
    throw new Error(`Hasura rejected the request (${response.status}):\n${JSON.stringify(parsed, null, 2)}`)
  }
  return parsed
}

async function main() {
  const checkOnly = process.argv.includes('--check')

  if (checkOnly) {
    const current = await callMetadataApi({ type: 'export_metadata', args: {} })
    const tables = current.sources?.[0]?.tables ?? []
    console.log(`${tables.length} tables tracked:`)
    for (const table of tables) {
      const roles = (table.select_permissions ?? []).map((p) => p.role).join(', ') || 'none'
      console.log(`  ${table.table.name.padEnd(22)} select: ${roles}`)
    }
    return
  }

  console.log(`applying metadata to ${ENDPOINT} …`)
  await callMetadataApi({
    type: 'replace_metadata',
    args: {
      // ⚠️ `allow_inconsistent_metadata: false`. The default is to accept metadata that references
      // a table which does not exist and mark it "inconsistent" — so a typo'd column name applies
      // cleanly and the permission it belonged to is simply not enforced. Failing instead means a
      // mistake here is a failed script, not a silently missing rule.
      allow_inconsistent_metadata: false,
      metadata,
    },
  })

  const applied = await callMetadataApi({ type: 'export_metadata', args: {} })
  const tables = applied.sources?.[0]?.tables ?? []

  console.log(`\ntracked ${tables.length} tables:`)
  for (const table of tables) {
    const selects = table.select_permissions ?? []
    const writes =
      (table.insert_permissions?.length ?? 0) +
      (table.update_permissions?.length ?? 0) +
      (table.delete_permissions?.length ?? 0)

    // ⚠️ This assertion is the point of printing anything. Every write in this app goes through
    // NestJS; a write permission appearing here means someone opened a second path to the tables
    // that skips every business rule. It should never be anything but 0.
    const flag = writes > 0 ? `  ⚠️  ${writes} WRITE PERMISSIONS` : ''
    console.log(`  ${table.table.name.padEnd(22)} select: ${selects.map((p) => p.role).join(', ') || 'none'}${flag}`)
  }

  const totalWrites = tables.reduce(
    (sum, t) =>
      sum +
      (t.insert_permissions?.length ?? 0) +
      (t.update_permissions?.length ?? 0) +
      (t.delete_permissions?.length ?? 0),
    0,
  )

  if (totalWrites > 0) {
    console.error(`\n⚠️  ${totalWrites} write permissions are configured. Every write belongs in NestJS.`)
    process.exit(1)
  }

  console.log('\ndone — reads only, as intended.')
}

main().catch((error) => {
  console.error(`\n${error.message}`)
  // Non-zero, so a shell script or CI step that chains commands actually stops here.
  process.exit(1)
})
